### PR TITLE
Setup Sentry instrumentation on top of the existing Signposter

### PR DESCRIPTION
**Pull request by stefanceriu**

### CODE REVIEW

File: ElementX/Sources/Services/BugReport/BugReportService.swift [hunk 1]
```diff
@@ -22,23 +22,21 @@ import UIKit
...
-    private let sentryURL: URL
     private let applicationId: String
     private let sdkGitSHA: String
     private let maxUploadSize: Int
     private let session: URLSession
-    private var lastCrashEventId: String?
     private let progressSubject = PassthroughSubject<Double, Never>()
     private var cancellables = Set<AnyCancellable>()
     
+    var lastCrashEventId: String?
```
Comment (pixlwave) [inline ElementX/Sources/Services/BugReport/BugReportService.swift:32]: 😉
```suggestion
    var lastCrashEventID: String?
```

File: ElementX/Sources/Services/Analytics/AnalyticsService.swift [hunk 2]
```diff
@@ -49,6 +49,7 @@ class AnalyticsService {
     init(client: AnalyticsClientProtocol, appSettings: AppSettings) {
         self.client = client
         self.appSettings = appSettings
+        signpost = Signposter(isDevelopmentBuild: AppSettings.isDevelopmentBuild)
```
Comment (pixlwave) [inline ElementX/Sources/Services/Analytics/AnalyticsService.swift:52]: Couldn't this init have been left on the variable itself?

File: ElementX/Sources/Screens/HomeScreen/View/HomeScreen.swift [hunk 3]
```diff
@@ -75,6 +75,7 @@ struct HomeScreen: View {
                     gradientView.alpha = 1
                 }
             }
+            .sentryTrace("HomeScreen")
```
Comment (pixlwave) [inline ElementX/Sources/Screens/HomeScreen/View/HomeScreen.swift:78]: Unimportant, but should we define these somewhere together so we don't accidentally use the same one twice? Or use `"\(Self.self)"` assuming that works on a view.

File: ElementX/Sources/Application/AppCoordinator.swift [hunk 4]
```diff
@@ -733,6 +733,51 @@ class AppCoordinator: AppCoordinatorProtocol, AuthenticationFlowCoordinatorDeleg
         }
     }
     
+    private func setupSentry() {
+        SentrySDK.start { options in
+            #if DEBUG
+            options.enabled = false
+            #endif
+            
+            options.dsn = ServiceLocator.shared.settings.bugReportSentryURL.absoluteString
```
Comment (pixlwave) [inline ElementX/Sources/Application/AppCoordinator.swift:742]: I think we have a local property for this one.
```suggestion
            options.dsn = appSettings.bugReportSentryURL.absoluteString
```

File: ElementX/Sources/Application/AppCoordinator.swift [hunk 5]
```diff
@@ -733,6 +733,51 @@ class AppCoordinator: AppCoordinatorProtocol, AuthenticationFlowCoordinatorDeleg
...
+            } else {
+                options.tracesSampleRate = 0.5
+            }
+            
+            // This callback is only executed once during the entire run of the program to avoid
+            // multiple callbacks if there are multiple crash events to send (see method documentation)
+            options.onCrashedLastRun = { event in
+                MXLog.error("Sentry detected a crash in the previous run: \(event.eventId.sentryIdString)")
+                ServiceLocator.shared.bugReportService.lastCrashEventId = event.eventId.sentryIdString
+            }
```
Comment (pixlwave) [inline ElementX/Sources/Application/AppCoordinator.swift:777]: Might be nice to have a `Sentry Started` log line here so we have an idea if a rageshake has it enabled or not.

File: ElementX/Sources/Services/Analytics/Signposter.swift [hunk 6]
```diff
@@ -49,37 +65,56 @@ class Signposter {
...
     
     func beginFirstRooms() {
         firstRoomsState = signposter.beginInterval(Name.firstRooms)
+        firstRoomsSpan = appStartupSpan.startChild(operation: "\(Name.firstRooms)")
     }
     
     func endFirstRooms() {
+        defer {
+            appStartupSpan.finish()
+        }
```
Comment (pixlwave) [inline ElementX/Sources/Services/Analytics/Signposter.swift:109]: Does this mean all the other spans that are children are stopped if they happened to still be running?

**Review by pixlwave — APPROVED** [review]
Very nice! I think my only question is should we perhaps hash the server name instead of collecting it?

File: ElementX/Sources/Services/Analytics/AnalyticsService.swift (hunk 2 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Analytics/AnalyticsService.swift:52]: Yeah, i initialy forgot I made that a static var

File: ElementX/Sources/Application/AppCoordinator.swift (hunk 4 above)
Comment (stefanceriu) [reply ElementX/Sources/Application/AppCoordinator.swift:742]: This way I don't need to access self in the block anymore

File: ElementX/Sources/Screens/HomeScreen/View/HomeScreen.swift (hunk 3 above)
Comment (stefanceriu) [reply ElementX/Sources/Screens/HomeScreen/View/HomeScreen.swift:78]: Yep okay, let's go with Self then

File: ElementX/Sources/Services/Analytics/Signposter.swift (hunk 6 above)
Comment (stefanceriu) [reply ElementX/Sources/Services/Analytics/Signposter.swift:109]: I would hope so but no idea honestly ¯\_(ツ)_/¯

**Pull request merged**